pypush/apns/lifecycle.py

```diff
@@ -89,53 +112,153 @@ async def reconnect(self):
+    @asynccontextmanager
+    async def _receive_stream(
+        self,
+        filter: filters.Filter[protocol.Command, T] = lambda c: c,
+        backlog: bool = True,
+    ):
+        async with self._broadcast.open_stream(backlog) as stream:
+            yield _util.FilteredStream(stream, filter)
+
+    async def _receive(
+        self, filter: filters.Filter[protocol.Command, T], backlog: bool = True
+    ):
+        async with self._receive_stream(filter, backlog) as stream:
             async for command in stream:
-                if isinstance(command, filter):
-                    yield command
-                    max -= 1
-                if max == 0:
-                    break
-
-    async def receive(self, filter: typing.Type[T]) -> T:
-        async for command in self.receive_stream(filter, 1):
-            return command
-        raise ValueError("No matching command received")
+                return command
+        raise ValueError("Did not receive expected command")
 
-    async def send(self, command: protocol.Command):
+    async def _send(self, command: protocol.Command):
         try:
-            assert self._conn is not None
-            await self._conn.send(command)
-        except Exception as e:
-            logging.warning(f"Error sending command, reconnecting")
+            async with self._send_lock:
+                assert self._conn is not None
+                await self._conn.send(command)
+        except Exception:
+            logging.warning("Error sending command, reconnecting")
             await self.reconnect()
-            await self.send(command)
+            await self._send(command)
+
+    async def _update_filter(self):
+        assert self.base_token is not None
+        await self._send(
+            protocol.FilterCommand(
+                token=self.base_token,
+                enabled_topic_hashes=[
+                    sha1(topic.encode()).digest() for topic in self._filters
+                ],
+            )
+        )
+
+    @asynccontextmanager
+    async def _filter(self, topics: list[str]):
+        assert self.base_token is not None
+        for topic in topics:
+            self._filters[topic] = self._filters.get(topic, 0) + 1
+        await self._update_filter()
+        yield
+        for topic in topics:
+            self._filters[topic] -= 1
+            if self._filters[topic] == 0:
+                del self._filters[topic]
+        await self._update_filter()
+
+    async def mint_scoped_token(self, topic: str) -> bytes:
+        topic_hash = sha1(topic.encode()).digest()
+        assert self.base_token is not None
+        await self._send(
+            protocol.ScopedTokenCommand(token=self.base_token, topic=topic_hash)
+        )
+        ack = await self._receive(filters.cmd(protocol.ScopedTokenAck))
+        assert ack.status == 0
+        return ack.scoped_token
+
+    @asynccontextmanager
+    async def notification_stream(
```

Comment:
Make sure to document that this does NOT ACK the notifications automatically, unlike `expect_notification`